File: regex.py
#!/usr/bin/env python3
"""
Implementación extendida de expresiones regulares en Python.
Soporta: ( | ) * + ? . \\d \\s \\w \\b [clase] [^clase]

Compila a un AFND y simula el AFND usando el algoritmo de Thompson.
Apto para procesar patrones como los que se encuentran en main.py.
"""
import sys

# Definiciones de clases para predicados de coincidencia de caracteres

class LiteralMatcher:
    def __init__(self, char):
        self.char = char
    def __call__(self, char, ignore_case=False):
        if ignore_case:
            return self.char.lower() == char.lower()
        return self.char == char
    def __repr__(self):
        return self.char

class DigitMatcher:
    def __call__(self, char):
        return char.isdigit()
    def __repr__(self):
        return r'\d'

class WhitespaceMatcher:
    def __call__(self, char):
        return char.isspace()
    def __repr__(self):
        return r'\s'

class WordMatcher:
    def __call__(self, char):
        return char.isalnum() or char == '_'
    def __repr__(self):
        return r'\w'

class DotMatcher:
    def __call__(self, char):
        return char != '\n'
    def __repr__(self):
        return '.'

class WordBoundaryAssertion:
    def __repr__(self):
        return r'\b'

class CharClassMatcher:
    def __init__(self, class_str):
        self.class_str = class_str
        self.negated = False
        if class_str.startswith('^'):
            self.negated = True
            class_str = class_str[1:]
        
        self.ranges = []
        self.chars = set()
        
        i = 0
        while i < len(class_str):
            if i + 2 < len(class_str) and class_str[i+1] == '-':
                start = class_str[i]
                end = class_str[i+2]
                self.ranges.append((start, end))
                i += 3
            else:
                self.chars.add(class_str[i])
                i += 1
                
    def __call__(self, char, ignore_case=False):
        c_val = char.lower() if ignore_case else char
        in_class = False
        
        chars_set = {c.lower() for c in self.chars} if ignore_case else self.chars
        if c_val in chars_set:
            in_class = True
        else:
            for start, end in self.ranges:
                s_val = start.lower() if ignore_case else start
                e_val = end.lower() if ignore_case else end
                if s_val <= c_val <= e_val:
                    in_class = True
                    break
        return not in_class if self.negated else in_class

    def __repr__(self):
        return f"[{self.class_str}]"


def tokenize_regex(re_str):
    """
    Parsea una cadena de expresión regular en átomos infijos.
    """
    atoms = []
    i = 0
    while i < len(re_str):
        char = re_str[i]
        if char == '\\':
            if i + 1 >= len(re_str):
                return None
            next_char = re_str[i+1]
            if next_char == 'd':
                atoms.append(DigitMatcher())
            elif next_char == 's':
                atoms.append(WhitespaceMatcher())
            elif next_char == 'w':
                atoms.append(WordMatcher())
            elif next_char == 'b':
                atoms.append(WordBoundaryAssertion())
            else:
                atoms.append(LiteralMatcher(next_char))  # Literal escapado
            i += 2
        elif char == '[':
            start_idx = i + 1
            end_idx = re_str.find(']', start_idx)
            if end_idx == -1:
                return None
            class_content = re_str[start_idx:end_idx]
            atoms.append(CharClassMatcher(class_content))
            i = end_idx + 1
        elif char == '.':
            atoms.append(DotMatcher())
            i += 1
        elif char in ('(', ')', '|', '*', '+', '?'):
            atoms.append(char)
            i += 1
        else:
            atoms.append(LiteralMatcher(char))
            i += 1
    return atoms


def re2post(infix_atoms):
    """
    Convierte átomos infijos a notación posfija.
    Inserta '.' como operador de concatenación explícito.
    """
    if infix_atoms is None:
        return None
    
    nalt = 0
    natom = 0
    buf = []
    paren = []
    
    for atom in infix_atoms:
        if atom == '(':
            if natom > 1:
                natom -= 1
                buf.append('.')
            paren.append({'nalt': nalt, 'natom': natom})
            nalt = 0
            natom = 0
        elif atom == '|':
            if natom == 0:
                return None
            while natom > 1:
                natom -= 1
                buf.append('.')
            natom = 0
            nalt += 1
        elif atom == ')':
            if not paren:
                return None
            if natom == 0:
                return None
            while natom > 1:
                natom -= 1
                buf.append('.')
            natom = 0
            while nalt > 0:
                buf.append('|')
                nalt -= 1
            p = paren.pop()
            nalt = p['nalt']
            natom = p['natom'] + 1
        elif atom in ('*', '+', '?'):
            if natom == 0:
                return None
            buf.append(atom)
        else:
            if natom > 1:
                natom -= 1
                buf.append('.')
            buf.append(atom)
            natom += 1
            
    if paren:
        return None
    while natom > 1:
        natom -= 1
        buf.append('.')
    while nalt > 0:
        buf.append('|')
        nalt -= 1
        
    return buf


class State:
    Match = 'Match'
    Split = 'Split'
    WordBoundary = 'WordBoundary'

    def __init__(self, c, out=None, out1=None):
        self.c = c
        self.out = out
        self.out1 = out1
        self.lastlist = 0

# El único estado global de coincidencia
matchstate = State(State.Match)


class Frag:
    def __init__(self, start, out_list):
        self.start = start
        self.out = out_list


def patch(out_list, s):
    for state, attr in out_list:
        setattr(state, attr, s)


def post2nfa(postfix_atoms):
    """
    Convierte átomos posfijos a AFND.
    """
    if postfix_atoms is None:
        return None
        
    stack = []
    for atom in postfix_atoms:
        if atom == '.':
            # Concatenar
            e2 = stack.pop()
            e1 = stack.pop()
            patch(e1.out, e2.start)
            stack.append(Frag(e1.start, e2.out))
        elif atom == '|':
            # Alternar
            e2 = stack.pop()
            e1 = stack.pop()
            s = State(State.Split, e1.start, e2.start)
            stack.append(Frag(s, e1.out + e2.out))
        elif atom == '?':
            # Cero o uno
            e = stack.pop()
            s = State(State.Split, e.start, None)
            stack.append(Frag(s, e.out + [(s, 'out1')]))
        elif atom == '*':
            # Cero o más
            e = stack.pop()
            s = State(State.Split, e.start, None)
            patch(e.out, s)
            stack.append(Frag(s, [(s, 'out1')]))
        elif atom == '+':
            # Uno o más
            e = stack.pop()
            s = State(State.Split, e.start, None)
            patch(e.out, s)
            stack.append(Frag(e.start, [(s, 'out1')]))
        elif isinstance(atom, WordBoundaryAssertion):
            s = State(State.WordBoundary)
            stack.append(Frag(s, [(s, 'out')]))
        else:
            s = State(atom)
            stack.append(Frag(s, [(s, 'out')]))

    if not stack:
        return None
    e = stack.pop()
    if stack:
        return None
        
    patch(e.out, matchstate)
    return e.start


class List:
    def __init__(self):
        self.s = []

listid = 0
l1 = List()
l2 = List()


def is_word_char(char):
    return char.isalnum() or char == '_'


def addstate(lst, s, string, index):
    """
    Añade un estado a la lista, siguiendo bifurcaciones y evaluando límites de palabra.
    """
    if s is None or s.lastlist == listid:
        return
    s.lastlist = listid
    if s.c == State.Split:
        addstate(lst, s.out, string, index)
        addstate(lst, s.out1, string, index)
        return
    if s.c == State.WordBoundary:
        prev_is_word = False
        if index > 0:
            prev_is_word = is_word_char(string[index-1])
            
        curr_is_word = False
        if index < len(string):
            curr_is_word = is_word_char(string[index])
            
        if prev_is_word != curr_is_word:
            addstate(lst, s.out, string, index)
        return
    lst.s.append(s)


def startlist(start, lst, string):
    """
    Calcula la lista de estados iniciales.
    """
    global listid
    lst.s = []
    listid += 1
    addstate(lst, start, string, 0)
    return lst


def ismatch(lst):
    return any(s is matchstate for s in lst.s)


def step(clist, char, nlist, string, index, ignore_case=False):
    """
    Avanza el AFND un carácter en el índice especificado hacia nlist.
    """
    global listid
    listid += 1
    nlist.s = []
    for s in clist.s:
        match = False
        if callable(s.c):
            try:
                match = s.c(char, ignore_case=ignore_case)
            except TypeError:
                match = s.c(char)
        elif isinstance(s.c, str):
            if ignore_case:
                match = s.c.lower() == char.lower()
            else:
                match = s.c == char
            
        if match:
            addstate(nlist, s.out, string, index + 1)


def match_nfa(start, string, ignore_case=False):
    """
    Ejecuta la coincidencia usando la simulación del AFND.
    """
    clist = startlist(start, l1, string)
    nlist = l2
    
    for index, char in enumerate(string):
        step(clist, char, nlist, string, index, ignore_case=ignore_case)
        clist, nlist = nlist, clist
        
    return ismatch(clist)


def match_longest_prefix(start, string, ignore_case=False):
    """
    Busca la longitud del prefijo más largo de string que coincide con el AFND.
    Devuelve -1 si ningún prefijo coincide.
    """
    clist = startlist(start, l1, string)
    nlist = l2
    
    longest_match_len = -1
    if ismatch(clist):
        longest_match_len = 0
        
    for index, char in enumerate(string):
        step(clist, char, nlist, string, index, ignore_case=ignore_case)
        clist, nlist = nlist, clist
        if len(clist.s) == 0:
            break
        if ismatch(clist):
            longest_match_len = index + 1
            
    return longest_match_len


class DState:
    """
    Representa un estado AFD: una lista de estados AFND almacenada en caché.
    """
    def __init__(self, nfa_list):
        self.l = nfa_list
        self.next = {}
        self.left = None
        self.right = None


alldstates = None


def listcmp(l1_states, l2_states):
    if len(l1_states) < len(l2_states):
        return -1
    if len(l1_states) > len(l2_states):
        return 1
    for s1, s2 in zip(l1_states, l2_states):
        if id(s1) < id(s2):
            return -1
        elif id(s1) > id(s2):
            return 1
    return 0


def dstate(lst):
    """
    Devuelve el DState en caché para la lista lst, creando uno nuevo si es necesario.
    """
    global alldstates
    sorted_states = sorted(lst.s, key=id)
    
    if alldstates is None:
        alldstates = DState(sorted_states)
        return alldstates
        
    dp = alldstates
    while True:
        cmp = listcmp(sorted_states, dp.l)
        if cmp < 0:
            if dp.left is None:
                dp.left = DState(sorted_states)
                return dp.left
            dp = dp.left
        elif cmp > 0:
            if dp.right is None:
                dp.right = DState(sorted_states)
                return dp.right
            dp = dp.right
        else:
            return dp


def startdstate(start, string):
    return dstate(startlist(start, l1, string))


def nextstate(d, char, string, index, ignore_case=False):
    clist = List()
    clist.s = d.l
    step(clist, char, l1, string, index, ignore_case=ignore_case)
    d.next[char] = dstate(l1)
    return d.next[char]


def match_dfa(start, string, ignore_case=False):
    """
    Ejecuta la simulación del AFD para determinar si coincide con la cadena.
    """
    d = start
    for index, char in enumerate(string):
        if char not in d.next:
            nextstate(d, char, string, index, ignore_case=ignore_case)
        d = d.next[char]
        
    return any(s is matchstate for s in d.l)


def has_boundary(postfix):
    return any(isinstance(atom, WordBoundaryAssertion) for atom in postfix)


def main():
    if len(sys.argv) < 3:
        print("uso: python regex.py [-n | -d] expr_regular cadena...", file=sys.stderr)
        print("opciones:\n  -n   usar simulación AFND (por defecto)\n  -d   usar caché AFD", file=sys.stderr)
        sys.exit(1)

    use_dfa = False
    arg_start = 2
    
    if sys.argv[1] == '-d':
        use_dfa = True
        if len(sys.argv) < 4:
            print("uso: python regex.py [-n | -d] expr_regular cadena...", file=sys.stderr)
            sys.exit(1)
        regexp_str = sys.argv[2]
        arg_start = 3
    elif sys.argv[1] == '-n':
        use_dfa = False
        if len(sys.argv) < 4:
            print("uso: python regex.py [-n | -d] expr_regular cadena...", file=sys.stderr)
            sys.exit(1)
        regexp_str = sys.argv[2]
        arg_start = 3
    else:
        regexp_str = sys.argv[1]

    infix = tokenize_regex(regexp_str)
    if infix is None:
        print(f"error al parsear la expresión regular: {regexp_str}", file=sys.stderr)
        sys.exit(1)

    post = re2post(infix)
    if post is None:
        print(f"estructura de expresión regular incorrecta: {regexp_str}", file=sys.stderr)
        sys.exit(1)

    # La simulación AFD no soporta aserciones de límites de forma contextual
    if use_dfa and has_boundary(post):
        print("aviso: se detectó la aserción de límite \\b; cambiando a simulación AFND", file=sys.stderr)
        use_dfa = False

    start = post2nfa(post)
    if start is None:
        print(f"error en la generación del AFND", file=sys.stderr)
        sys.exit(1)

    if use_dfa:
        for i in range(arg_start, len(sys.argv)):
            # Recrea el estado inicial para diferentes cadenas para manejar el contexto correctamente
            dstart_i = startdstate(start, sys.argv[i])
            if match_dfa(dstart_i, sys.argv[i]):
                print(sys.argv[i])
    else:
        for i in range(arg_start, len(sys.argv)):
            if match_nfa(start, sys.argv[i]):
                print(sys.argv[i])

if __name__ == '__main__':
    main()

File: regex.c
/*
 * Implementación extendida de expresiones regulares en C.
 * Soporta: ( | ) * + ? . \\d \\s \\w \\b [clase] [^clase]
 *
 * Compila a un AFND y simula el AFND usando el algoritmo de Thompson,
 * con almacenamiento en caché opcional para AFD para expresiones sin límites de palabra.
 * 
 * Copyright (c) 2007 Russ Cox.
 * Se puede distribuir bajo la licencia MIT, ver al final del archivo.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * Representa las clases de coincidencia de estados del AFND
 */
enum
{
	Match = 256,
	Split = 257,
	Any = 258,        /* . */
	Digit = 259,      /* \d */
	Whitespace = 260, /* \s */
	Word = 261,       /* \w */
	Boundary = 262,   /* \b */
	CharClass = 263,  /* [...] */
	Concat = 264      /* Operador de concatenación */
};

typedef struct CharClassData CharClassData;
struct CharClassData
{
	int negated;
	char chars[256];
	int nchars;
	struct {
		char start;
		char end;
	} ranges[128];
	int nranges;
};

typedef struct State State;
struct State
{
	int c;
	State *out;
	State *out1;
	int lastlist;
	CharClassData *cc; /* Se usa solo si c == CharClass */
};

State matchstate = { Match };	/* estado de coincidencia */
int nstate;

/* Asigna e inicializa un Estado */
State*
state(int c, State *out, State *out1)
{
	State *s;
	
	nstate++;
	s = malloc(sizeof *s);
	s->lastlist = 0;
	s->c = c;
	s->out = out;
	s->out1 = out1;
	s->cc = NULL;
	return s;
}

/*
 * Un fragmento de AFND parcialmente construido sin el estado de coincidencia completado.
 */
typedef struct Frag Frag;
typedef union Ptrlist Ptrlist;
struct Frag
{
	State *start;
	Ptrlist *out;
};

Frag
frag(State *start, Ptrlist *out)
{
	Frag n = { start, out };
	return n;
}

union Ptrlist
{
	Ptrlist *next;
	State *s;
};

/* Crea una lista con un único elemento que contiene outp. */
Ptrlist*
list1(State **outp)
{
	Ptrlist *l;
	
	l = (Ptrlist*)outp;
	l->next = NULL;
	return l;
}

/* Conecta la lista de estados en 'out' para que apunten a 'start'. */
void
patch(Ptrlist *l, State *s)
{
	Ptrlist *next;
	
	for(; l; l=next){
		next = l->next;
		l->s = s;
	}
}

/* Une las dos listas l1 y l2, devolviendo la combinación. */
Ptrlist*
append(Ptrlist *l1, Ptrlist *l2)
{
	Ptrlist *oldl1;
	
	oldl1 = l1;
	while(l1->next)
		l1 = l1->next;
	l1->next = l2;
	return oldl1;
}

typedef struct Atom Atom;
struct Atom
{
	int type;
	CharClassData *cc;
};

/* Parsea una cadena de expresión regular en átomos infijos */
int
tokenize_regex(char *re, Atom *atoms, int max_atoms)
{
	int n = 0;
	char *p = re;
	while(*p) {
		if(n >= max_atoms)
			return -1;
		if(*p == '\\') {
			p++;
			if(*p == '\0')
				return -1;
			if(*p == 'd') {
				atoms[n].type = Digit;
				atoms[n].cc = NULL;
			} else if(*p == 's') {
				atoms[n].type = Whitespace;
				atoms[n].cc = NULL;
			} else if(*p == 'w') {
				atoms[n].type = Word;
				atoms[n].cc = NULL;
			} else if(*p == 'b') {
				atoms[n].type = Boundary;
				atoms[n].cc = NULL;
			} else {
				atoms[n].type = *p;
				atoms[n].cc = NULL;
			}
			p++;
		} else if(*p == '[') {
			p++;
			char *start = p;
			while(*p && *p != ']')
				p++;
			if(*p == '\0')
				return -1;
			CharClassData *cc = malloc(sizeof *cc);
			memset(cc, 0, sizeof *cc);
			if(*start == '^') {
				cc->negated = 1;
				start++;
			}
			char *ccp = start;
			while(ccp < p) {
				if(ccp + 2 < p && ccp[1] == '-') {
					cc->ranges[cc->nranges].start = ccp[0];
					cc->ranges[cc->nranges].end = ccp[2];
					cc->nranges++;
					ccp += 3;
				} else {
					cc->chars[cc->nchars++] = *ccp;
					ccp++;
				}
			}
			atoms[n].type = CharClass;
			atoms[n].cc = cc;
			p++; /* omitir ']' */
		} else if(*p == '.') {
			atoms[n].type = Any;
			atoms[n].cc = NULL;
			p++;
		} else {
			atoms[n].type = *p;
			atoms[n].cc = NULL;
			p++;
		}
		n++;
	}
	return n;
}

/* Convierte átomos infijos a notación posfija */
int
re2post_atoms(Atom *infix, int ninfix, Atom *postfix)
{
	int nalt, natom;
	int i;
	int nout = 0;
	struct {
		int nalt;
		int natom;
	} paren[100], *p;
	
	p = paren;
	nalt = 0;
	natom = 0;
	
	for(i = 0; i < ninfix; i++) {
		Atom atom = infix[i];
		switch(atom.type) {
		case '(':
			if(natom > 1) {
				natom--;
				postfix[nout++].type = Concat;
				postfix[nout-1].cc = NULL;
			}
			if(p >= paren + 100)
				return -1;
			p->nalt = nalt;
			p->natom = natom;
			p++;
			nalt = 0;
			natom = 0;
			break;
		case '|':
			if(natom == 0)
				return -1;
			while(natom > 1) {
				natom--;
				postfix[nout++].type = Concat;
				postfix[nout-1].cc = NULL;
			}
			natom = 0;
			nalt++;
			break;
		case ')':
			if(p == paren)
				return -1;
			if(natom == 0)
				return -1;
			while(natom > 1) {
				natom--;
				postfix[nout++].type = Concat;
				postfix[nout-1].cc = NULL;
			}
			natom = 0;
			while(nalt > 0) {
				postfix[nout++].type = '|';
				postfix[nout-1].cc = NULL;
				nalt--;
			}
			p--;
			nalt = p->nalt;
			natom = p->natom;
			natom++;
			break;
		case '*':
		case '+':
		case '?':
			if(natom == 0)
				return -1;
			postfix[nout++] = atom;
			break;
		default:
			if(natom > 1) {
				natom--;
				postfix[nout++].type = Concat;
				postfix[nout-1].cc = NULL;
			}
			postfix[nout++] = atom;
			natom++;
			break;
		}
	}
	if(p != paren)
		return -1;
	while(natom > 1) {
		natom--;
		postfix[nout++].type = Concat;
		postfix[nout-1].cc = NULL;
	}
	while(nalt > 0) {
		postfix[nout++].type = '|';
		postfix[nout-1].cc = NULL;
		nalt--;
	}
	return nout;
}

/* Convierte átomos posfijos a AFND. Devuelve el estado inicial */
State*
post2nfa_atoms(Atom *postfix, int npostfix)
{
	int i;
	Frag stack[1000], *stackp, e1, e2, e;
	State *s;

	#define push(s) *stackp++ = s
	#define pop() *--stackp

	stackp = stack;
	for(i=0; i<npostfix; i++){
		Atom atom = postfix[i];
		switch(atom.type){
		default:
			s = state(atom.type, NULL, NULL);
			if(atom.type == CharClass) {
				s->cc = atom.cc;
			}
			push(frag(s, list1(&s->out)));
			break;
		case Concat:	/* concatenar */
			e2 = pop();
			e1 = pop();
			patch(e1.out, e2.start);
			push(frag(e1.start, e2.out));
			break;
		case '|':	/* alternar */
			e2 = pop();
			e1 = pop();
			s = state(Split, e1.start, e2.start);
			push(frag(s, append(e1.out, e2.out)));
			break;
		case '?':	/* cero o uno */
			e = pop();
			s = state(Split, e.start, NULL);
			push(frag(s, append(e.out, list1(&s->out1))));
			break;
		case '*':	/* cero o más */
			e = pop();
			s = state(Split, e.start, NULL);
			patch(e.out, s);
			push(frag(s, list1(&s->out1)));
			break;
		case '+':	/* uno o más */
			e = pop();
			s = state(Split, e.start, NULL);
			patch(e.out, s);
			push(frag(e.start, list1(&s->out1)));
			break;
		}
	}

	e = pop();
	if(stackp != stack)
		return NULL;

	patch(e.out, &matchstate);
	return e.start;
#undef pop
#undef push
}

typedef struct List List;
struct List
{
	State **s;
	int n;
};
List l1, l2;
static int listid;

int iswordchar(int c)
{
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_';
}

void addstate(List*, State*, char*, char*);
void step(List*, int, List*, char*, char*);

/* Calcula la lista de estados iniciales */
List*
startlist(State *start, List *l, char *str)
{
	l->n = 0;
	listid++;
	addstate(l, start, str, str);
	return l;
}

/* Comprueba si la lista de estados contiene una coincidencia. */
int
ismatch(List *l)
{
	int i;

	for(i=0; i<l->n; i++)
		if(l->s[i] == &matchstate)
			return 1;
	return 0;
}

/* Añade s a l, siguiendo bifurcaciones y límites de palabra */
void
addstate(List *l, State *s, char *str, char *p)
{
	if(s == NULL || s->lastlist == listid)
		return;
	s->lastlist = listid;
	if(s->c == Split){
		addstate(l, s->out, str, p);
		addstate(l, s->out1, str, p);
		return;
	}
	if(s->c == Boundary){
		int prev_is_word = 0;
		int curr_is_word = 0;
		if(p > str)
			prev_is_word = iswordchar(*(p-1));
		if(*p != '\0')
			curr_is_word = iswordchar(*p);
		if(prev_is_word != curr_is_word)
			addstate(l, s->out, str, p);
		return;
	}
	l->s[l->n++] = s;
}

/* Comprueba si el estado s coincide con el carácter c */
int
matchchar(State *s, int c)
{
	if(s->c < 256)
		return s->c == c;
	switch(s->c){
	case Any:
		return c != '\n';
	case Digit:
		return c >= '0' && c <= '9';
	case Whitespace:
		return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\v' || c == '\f';
	case Word:
		return iswordchar(c);
	case CharClass:
		{
			CharClassData *cc = s->cc;
			int in_class = 0;
			int i;
			for(i=0; i<cc->nchars; i++) {
				if(cc->chars[i] == c) {
					in_class = 1;
					break;
				}
			}
			if(!in_class) {
				for(i=0; i<cc->nranges; i++) {
					if(c >= cc->ranges[i].start && c <= cc->ranges[i].end) {
						in_class = 1;
						break;
					}
				}
			}
			return cc->negated ? !in_class : in_class;
		}
	}
	return 0;
}

/*
 * Avanza el AFND un paso desde clist pasando el carácter c en el puntero p,
 * para crear el siguiente conjunto de estados nlist.
 */
void
step(List *clist, int c, List *nlist, char *str, char *p)
{
	int i;
	State *s;

	listid++;
	nlist->n = 0;
	for(i=0; i<clist->n; i++){
		s = clist->s[i];
		if(matchchar(s, c))
			addstate(nlist, s->out, str, p + 1);
	}
}

/* Ejecuta el AFND para determinar si coincide con str. */
int
match_nfa(State *start, char *str)
{
	int c;
	List *clist, *nlist, *t;
	char *p;

	clist = startlist(start, &l1, str);
	nlist = &l2;
	for(p=str; *p; p++){
		c = *p & 0xFF;
		step(clist, c, nlist, str, p);
		t = clist; clist = nlist; nlist = t;	/* intercambiar clist, nlist */
	}
	return ismatch(clist);
}

/*
 * Representa un estado AFD: una lista de estados AFND en caché.
 */
typedef struct DState DState;
struct DState
{
	List l;
	DState *next[256];
	DState *left;
	DState *right;
};

static int
listcmp(List *l1, List *l2)
{
	int i;

	if(l1->n < l2->n)
		return -1;
	if(l1->n > l2->n)
		return 1;
	for(i=0; i<l1->n; i++)
		if(l1->s[i] < l2->s[i])
			return -1;
		else if(l1->s[i] > l2->s[i])
			return 1;
	return 0;
}

static int
ptrcmp(const void *a, const void *b)
{
	State *sa = *(State**)a;
	State *sb = *(State**)b;
	if(sa < sb)
		return -1;
	if(sa > sb)
		return 1;
	return 0;
}

DState *alldstates;
DState*
dstate(List *l)
{
	int i;
	DState **dp, *d;

	qsort(l->s, l->n, sizeof l->s[0], ptrcmp);
	dp = &alldstates;
	while((d = *dp) != NULL){
		i = listcmp(l, &d->l);
		if(i < 0)
			dp = &d->left;
		else if(i > 0)
			dp = &d->right;
		else
			return d;
	}
	
	d = malloc(sizeof *d + l->n*sizeof l->s[0]);
	memset(d, 0, sizeof *d);
	d->l.s = (State**)(d+1);
	memmove(d->l.s, l->s, l->n*sizeof l->s[0]);
	d->l.n = l->n;
	*dp = d;
	return d;
}

DState*
startdstate(State *start, char *str)
{
	return dstate(startlist(start, &l1, str));
}

DState*
nextstate(DState *d, int c, char *str, char *p)
{
	step(&d->l, c, &l1, str, p);
	return d->next[c] = dstate(&l1);
}

int
match_dfa(DState *start, char *str)
{
	DState *d, *next;
	int c;
	char *p;
	
	d = start;
	for(p=str; *p; p++){
		c = *p & 0xFF;
		if((next = d->next[c]) == NULL)
			next = nextstate(d, c, str, p);
		d = next;
	}
	return ismatch(&d->l);
}

int
main(int argc, char **argv)
{
	int i;
	State *start;
	int use_dfa = 0;
	int arg_start = 2;
	char *regexp_str = NULL;
	Atom infix[1000];
	Atom post[1000];
	int ninfix, npost;

	if(argc < 3){
		fprintf(stderr, "uso: regex [-n | -d] expr_regular cadena...\n");
		fprintf(stderr, "opciones:\n  -n   usar simulación AFND (por defecto)\n  -d   usar caché AFD\n");
		return 1;
	}

	if(strcmp(argv[1], "-d") == 0) {
		use_dfa = 1;
		if (argc < 4) {
			fprintf(stderr, "uso: regex [-n | -d] expr_regular cadena...\n");
			return 1;
		}
		regexp_str = argv[2];
		arg_start = 3;
	} else if(strcmp(argv[1], "-n") == 0) {
		use_dfa = 0;
		if (argc < 4) {
			fprintf(stderr, "uso: regex [-n | -d] expr_regular cadena...\n");
			return 1;
		}
		regexp_str = argv[2];
		arg_start = 3;
	} else {
		regexp_str = argv[1];
	}
	
	ninfix = tokenize_regex(regexp_str, infix, 1000);
	if(ninfix < 0) {
		fprintf(stderr, "error al parsear la expresión regular: %s\n", regexp_str);
		return 1;
	}

	npost = re2post_atoms(infix, ninfix, post);
	if(npost < 0){
		fprintf(stderr, "bad regexp structure: %s\n", regexp_str);
		return 1;
	}

	/* Verifica si hay aserciones de límites de palabras; vuelve a simulación AFND si las hay */
	int has_boundary = 0;
	for(i=0; i<npost; i++) {
		if(post[i].type == Boundary) {
			has_boundary = 1;
			break;
		}
	}
	if(use_dfa && has_boundary) {
		fprintf(stderr, "aviso: se detectó la aserción de límite \\b; cambiando a simulación AFND\n");
		use_dfa = 0;
	}

	start = post2nfa_atoms(post, npost);
	if(start == NULL){
		fprintf(stderr, "error en la generación del AFND\n");
		return 1;
	}
	
	l1.s = malloc(nstate*sizeof l1.s[0]);
	l2.s = malloc(nstate*sizeof l2.s[0]);

	if (use_dfa) {
		for(i=arg_start; i<argc; i++) {
			DState *dstart_i = startdstate(start, argv[i]);
			if(match_dfa(dstart_i, argv[i]))
				printf("%s\n", argv[i]);
		}
	} else {
		for(i=arg_start; i<argc; i++) {
			if(match_nfa(start, argv[i]))
				printf("%s\n", argv[i]);
		}
	}

	return 0;
}

/*
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the
 * Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall
 * be included in all copies or substantial portions of the
 * Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
 * KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
